src/manual/devel/best-practices/microservices.md

```diff
@@ -0,0 +1,307 @@
+Its README describes how to set it up generally, but for a TaskCluster service, you will need to install `babel-compile` and `babel-preset-taskcluster`.
+Include the following in your `package.json`:
+
+```js
+  "scripts": {
+    "compile": "babel-compile -p taskcluster src:lib test:.test",
+    "pretest": "npm run compile",
+    "install": "npm run compile",
+  }
+```
+
+### Dependencies
+
+Try to keep up-to-date with the latest versions of all TaskCluster libraries.
+In general, the implications of updating these libraries should be clear, and the authors are easy to find when things go badly.
+
+Other dependencies should also be kept up-to-date as much as possible.
+Tools like [Greenkeeper](https://greenkeeper.io/) can be very useful for this purpose.
+
+### Shrinkwrap
+
+Dependencies for deployed services are handled with [`npm shrinkwrap`](https://docs.npmjs.com/cli/shrinkwrap).
+This is to ensure that we are deploying a consistent tree of package versions, and that dependency upgrades are controlled and not automatic or surprising.
+
+When creating a new service, run `npm shrinkwrap` to generate an `npm-shrinkwrap.json` based on the packages you have installed in your development environment, then check that file into the Git repository.
+When beginning development on an existing service, the usual `npm install` will install the specific versions named in the `npm-shrinkwrap.json`.
+
+*Important*: Do not use `devDependencies`, as npm versions before 3.0 do not deal with them correctly and will cause failures in deployment that do not appear in local runs or Travis.
+Instead, list all dependencies in `dependencies`.
+
+When changing a service's dependencies, the easiest approach is to use `npm install --save` to install the package and update `package.json` at the same time.
+Once this is complete, run `npm shrinkwrap` and add the modified `npm-shrinkwrap.json` to your commit.
+Pull requests with a modified `package.json` that do not have a corresponding `npm-shrinkwrap.json` diff should be viewed with suspicion!
+
+If is the service owner's responsibility to keep dependencies up to date.
+The `npm outdated` command gives a useful overview of available updates.
+In general, try to keep packages up to date within semver constraints (so, fix things displayed in red in `npm outdated`), but be cautious that the new code you are incorporating into your service is trustworthy.
+In an ideal world, that means a thorough security review.
+In the real world, that probably means a lot less.
+
+## Testing
+
+### Test Setup
+
+Use Mocha to run unit tests, in the `test/` directory.
+Include the following in `mocha.opts`, most critically the `--ui tdd`.
+
+```
+--ui tdd
+--timeout 30s
+--reporter spec
+```
+
+Name the test files `test/*_test.js`, so they will be matched by the `npm test` script given above.
+These files should require production code from `../lib/`: `foo = require('../lib/foo')`.
```

Comment:
We should add:
```
Because we transpile the tests to different directory than Mocha's default (`.test` instead of `test`), we need to be careful that the bash globbing will match all test files.  The easiest way to ensure this is to have all test files directly in the `test` directory and no sub-directories 
```

But, I'd rather that we use `test-src:test` and let the mocha discovery stuff do discovery :)